File: examples/register_pair_arithmetic.py
from emu_8085 import Data, Instruction, Machine, Mem, Opcode, Program


def main_run() -> None:
    # Initialize machine
    machine = Machine.create(address_lines=16, data_lines=8)
    cpu = machine.cpu

    # Initialize register pairs:
    # BC = 0x1234
    # DE = 0x000F
    # HL = 0x1000
    cpu.reg_b.write(Data.byte(0x12))
    cpu.reg_c.write(Data.byte(0x34))
    cpu.reg_d.write(Data.byte(0x00))
    cpu.reg_e.write(Data.byte(0x0F))
    cpu.reg_h.write(Data.byte(0x10))
    cpu.reg_l.write(Data.byte(0x00))

    program = Program([
        Instruction(Opcode.INX_BC),  # BC = 0x1235
        Instruction(Opcode.DCX_DE),  # DE = 0x000E
        Instruction(Opcode.DAD_BC),  # HL = HL + BC = 0x1000 + 0x1235 = 0x2235
        Instruction(Opcode.DAD_DE),  # HL = HL + DE = 0x2235 + 0x000E = 0x2243
        Instruction(Opcode.HLT)
    ])

    machine.load(program, Mem(0x0000))
    machine.run()

    bc_val = (cpu.reg_b.read().value << 8) | cpu.reg_c.read().value
    de_val = (cpu.reg_d.read().value << 8) | cpu.reg_e.read().value
    hl_val = (cpu.reg_h.read().value << 8) | cpu.reg_l.read().value

    print("16-Bit Register Pair Arithmetic Example State:")
    print(f"Register Pair BC: {hex(bc_val)} (Expected: 0x1235)")
    print(f"Register Pair DE: {hex(de_val)} (Expected: 0x000e)")
    print(f"Register Pair HL: {hex(hl_val)} (Expected: 0x2243)")
    print(f"Carry Flag (CY): {cpu.flag_reg.carry}")

if __name__ == "__main__":
    main_run()

File: examples/register_pair_arithmetic.rs
//! Ported from the reference project's `examples/register_pair_arithmetic.py`.
//! Run with `cargo run --example register_pair_arithmetic`.

use emu8085::{Addr, Instruction, Machine, Opcode, Program, Reg16};

fn main() {
    let mut machine = Machine::create(16, 8);
    let cpu = &mut machine.cpu;

    // Initialize register pairs:
    // BC = 0x1234
    // DE = 0x000F
    // HL = 0x1000
    cpu.regs.set16(Reg16::BC, Addr(0x1234));
    cpu.regs.set16(Reg16::DE, Addr(0x000F));
    cpu.regs.set16(Reg16::HL, Addr(0x1000));

    let program = Program::new(vec![
        Instruction::new(Opcode::INX_BC),  // BC = 0x1235
        Instruction::new(Opcode::DCX_DE),  // DE = 0x000E
        Instruction::new(Opcode::DAD_BC),  // HL = HL + BC = 0x1000 + 0x1235 = 0x2235
        Instruction::new(Opcode::DAD_DE),  // HL = HL + DE = 0x2235 + 0x000E = 0x2243
        Instruction::new(Opcode::HLT),
    ]);

    machine.load(&program, Addr(0x0000)).expect("program compiles");
    machine.run();

    let cpu = &machine.cpu;
    let bc_val = cpu.regs.get16(Reg16::BC);
    let de_val = cpu.regs.get16(Reg16::DE);
    let hl_val = cpu.regs.get16(Reg16::HL);

    println!("16-Bit Register Pair Arithmetic Example State:");
    println!("Register Pair BC: 0x{:04X} (Expected: 0x1235)", bc_val.0);
    println!("Register Pair DE: 0x{:04X} (Expected: 0x000E)", de_val.0);
    println!("Register Pair HL: 0x{:04X} (Expected: 0x2243)", hl_val.0);
    println!("Carry Flag (CY): {}", cpu.flags.carry);

    assert_eq!(bc_val.0, 0x1235);
    assert_eq!(de_val.0, 0x000E);
    assert_eq!(hl_val.0, 0x2243);
}
